WalletWasabi/Crypto/ZeroKnowledge/Challenge.cs

```diff
@@ -0,0 +1,42 @@
+using NBitcoin.Secp256k1;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using WalletWasabi.Crypto.Groups;
+using WalletWasabi.Helpers;
+
+namespace WalletWasabi.Crypto.ZeroKnowledge
+{
+	public static class Challenge
+	{
+		public static Scalar Build(GroupElement publicPoint, GroupElement nonce, IEnumerable<GroupElement> generators)
+		{
+			Guard.False($"{nameof(publicPoint)}.{nameof(publicPoint.IsInfinity)}", publicPoint.IsInfinity);
+			Guard.False($"{nameof(nonce)}.{nameof(nonce.IsInfinity)}", nonce.IsInfinity);
+			foreach (var generator in generators)
+			{
+				Guard.False($"{nameof(generator)}.{nameof(generator.IsInfinity)}", generator.IsInfinity);
+			}
+
+			if (publicPoint == nonce)
+			{
+				throw new InvalidOperationException($"{nameof(publicPoint)} and {nameof(nonce)} should not be equal.");
+			}
+
+			return HashToScalar(new[] { publicPoint, nonce }.Concat(generators));
```

Comment:
This `HashToScalar` is known as `Fiat-Shamir transformation` (or fiat-shamir heuristic but i don't really understand why heuristic make sense here). I think this could be a better name because the following should make sense to cryptographers i think.

```c#
FiatShamir(nonce /*or commitment*/, statementToProve)` 
```

Reply:
I suspect it's called a heuristic because strictly speaking it converts proofs of knowledge to arguments of knowledge? not sure... it's also called a Fiat-Shamir transform(ation) if i'm not mistaken

Reply:
trying to look up a bit more about this, i'm not convinced my statement was accurate... @seresistvanandras? my understanding was the difference between proofs of knowledge and arguments of knowledge is that the latter only ensure soundness against a computationally bounded adversary, but it appears the definition is more subtle? at any rate i still think it's a "heuristic" because it transforms a 3 message interactive proof system into a 1 message non interactive one, and this may have to do with the fact that even the security proofs in the ROM model only came after if i'm not mistaken... sorry this distraction is not really relevant to the code in any way but it's just something i've wondered about for a while myself

Reply:
You're right about the distinction between proofs of knowledge and argument of knowledge. I think the name heuristic is just a historic naming convention. Not sure if it has specific meaning in this context. I suppose It's a heuristic because it cannot be proven in the standard model, it only yields a secure protocol in the random oracle model, ie. if you assume that the hash function indeed behaves like a random function.

Reply:
So should I change it `FiatShamir`?

Reply:
IMO both are fine, it's more of a coding style question. Yet another option that is somewhere in between might be something like `ComputeChallenge`... I don't think it would make much of a difference to cryptographers or cryptographic engineers auditing the codebase, but it would to those with less background.

Calling it `FiatShamir` would be most helpful to readers who would benefit from knowing what background information to look up, if they would google something like that.

`HashToScalar` on the other hand would be most helpful to those who would benefit more from first understanding what's happenning in a mechanistic sense, and then learning the rationale behind it.

Both should be served by a doc comment on the function explaining it's to be used for the Fiat Shamir heuristic if it's called `HashToScalar` or mention that it works by the inputs if it's called `FiatShamir`.

Reply:
Added as a comment.